MauroDataMapper-NHSD/nhsd-datadictionary-orchestration
# gh-14 Update Orchestrator to support OpenID Connect authentication

Resolves #14 

# Overview

- Add Mauro features switch service, used to verify that the OpenID Connect feature is switched on
- Update sign in form to use Angular typed form
- Load and display the OpenID Connect providers available
- Update SecurityService to support OpenID Connect actions
- Added tests to verify SecurityService works with OpenID Connection functions
- Add static redirect page for OpenID Connect to return back to
- Continue Mauro authorization once OpenID Connect provider has returned back to us
- On successful login, redirect to the first page in the Orchestrator
- Ensure a user authenticated via OIDC is an administrator, otherwise redirect back to the home page

# Testing

Testing locally will require some configuration, talk to @pjmonks for full details.

## Mauro plugins

You will need the `uk.ac.ox.softeng.maurodatamapper.plugins:mdm-plugin-authentication-openid-connect:2.2.0` plugin included in `mdm-application-build`

## OpenID Connect provider

There is an OpenID Connect provider available on the OCC Azure tenant which can authenticate against an @oxfordcc.co.uk account. Set this up in Mauro as follows:

1. Sign in as the administrator
2. Go to the OpenID Connect admin page and add a new provider
3. Set label to "Microsoft Azure"
4. Client ID and client secret can be found in Devolutions under the "OUH" folder.
5. Tick "Use discovery document for endpoints" and use this discovery document URL: https://login.microsoftonline.com/bd91e5a9-7949-44a9-8c9b-b66c8c527d47/v2.0/.well-known/openid-configuration
6. Click "Create provider"

## Test OIDC on Mauro

Sign out as the administrator, then click "Log in" and click on the new "Microsoft Azure" button. This should redirect to Microsoft where you can authenticate against your @oxfordcc.co.uk account. This should then redirect you back to Mauro and viewing the model tree.

## Test OIDC on Orchestrator

Now got to the Orchestrator and click the "Sign in" button. A "Microsoft Azure" button should appear here too, which should function the same as above.

Verify that:

1. Initial sign in might not work, since your @oxfordcc.co.uk account is not in the "administrators" group - sign in should be cancelled, and you are redirected to the home page
2. Add your @oxfordcc.co.uk user account in Mauro to the "administrators" group
    1. Sign in to Mauro as the administrator account
    2. Go to the "Manage groups" admin page
    3. Edit the "administrators" group
    4. Add your @oxfordcc.co.uk user account 
    5. Sign out of Mauro
    6. Go back to Orchestrator and try again. You should eventually end up on the "Branches" page
